[REVIEW] Generalize rearrange_by_column_tasks and add DataFrame.shuffle

rjzamora opened the pull request:
Following the discussion in [#5741](https://github.com/dask/dask/pull/5741) (particularly in/after [this comment](https://github.com/dask/dask/pull/5741#issuecomment-608079985)), this draft PR generalizes the `rearrange_by_column_tasks` implementation to handle hash-based shuffling without the existence of a "_partitions" column (including support for multiple shuffle-index columns).

**TODO**:
- [x] Include benchmarking numbers for ~both cudf and~ pandas-backed operations

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

mrocklin commented:
It looks like hasn't received any review over thepast few weeks.  My apologies @rjzamora .  Is this still live ?

rjzamora commented:
> It looks like hasn't received any review over thepast few weeks. My apologies @rjzamora . Is this still live ?

Thanks for the ping @mrocklin - I have been meaning to revisit and revise this. 

Overall, it would be wonderful to have a generalized `rearrange_by_column_tasks` that will actually handle more than one column (and simply perform a hash in both `shuffle_group` and `shuffle_group_2` in lieu of using a `"_partitions"`) column.  I feel that dask_cudf is currently duplicating code just to avoid creating/assigning the `"_partitions"` column.  We would definitely prefer to use upstream-Dask for all shuffling.  Although the simplicity of the single-column routine is likely best for pandas-backed DataFrame shuffling,  we are really hoping to introduce a minimal amount of flexibility here.   Since device memory is a valuable resource, we are doing our best to shed every byte we can (to reduce the amount of data we need to spill to host memory).

It would also be nice to introduce a mechanism to use the above memory optimization in the merge code path.  However, I certainly understand if you would rather `dask_cudf` implement/maintain it's own wrapper to avoid `"_partition"` creation.

rjzamora commented:
Following the comments in #6133 , I decided to add the `DataFrame.shuffle` API in this PR. (cc @TomAugspurger )

In order to allow dask_cudf to minimize the cost of the `"_partitions"` column creation/assignment (for workflows with high memory pressure), I added the `shuffle_dtype` argument in a few places.  This argument allows the user to explicitly specify the type of the `"_partitions"` column.  If this argument is set to `False`, `"_partitions"` will *not* be created at all (and hashing will be performed within all local shuffle-group operations).

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1178,6 +1178,64 @@ def repartition(
 
+    def shuffle(
+        self,
+        index,
```
```suggestion
        on,
```

What do you think about this name? `index` is a bit overloaded.

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1178,6 +1178,64 @@ def repartition(
+        """ Rearrange DataFrame into new partitions by index
+
+        Uses hashing to map rows to output partitions. After this operation,
+        rows with the same index element(s) will be in the same partition.
```
It's implied, but maybe mention that the result will have unknown divisions?

TomAugspurger replied:
Ah, I see you mention that in the Notes. That's fine too.

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1178,6 +1178,64 @@ def repartition(
+
+        Parameters
+        ----------
+        index : str, list of str, or _Frame
```
```suggestion
        index : str, list of str, or Series, Index, or DataFrame
```

`_Frame` isn't public so best to keep it out of docstrings.

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1178,6 +1178,64 @@ def repartition(
+        --------
+        >>> df = df.shuffle(df.columns[0])  # doctest: +SKIP
+        """
+        from .shuffle import shuffle
```
The name here may be conflicting with the `shuffle` keyword :)

rjzamora commented:
Thanks for reviewing Tom!

> Is there any reason to not use shuffle_dtype=False when using hash-based partitioning?

There be cases in which it is more performant to create the `"_partitions"` column, but I haven't experienced this.  It seems likely that the "_partitions"-based workflow was used to align the shuffle algorithm to be used cleanly wit `set_index`.  In my experience it is a significant memory and communication benefit to avoid adding an additional column (hence my motivation for this PR).

Simple benchmark as motivation:

```python
from dask.distributed import LocalCluster, Client, wait
from dask.datasets import timeseries

cluster = LocalCluster(n_workers=8)
client = Client(cluster)
ddf = timeseries(start='2000-01-01', end='2000-12-31', partition_freq='1d')

# Default
%timeit wait(ddf.shuffle("id", shuffle="tasks", shuffle_dtype=None).persist())

# Minimal dtype
%timeit wait(ddf.shuffle("id", shuffle="tasks", shuffle_dtype="uint16").persist())

# No "_partitions"
%timeit wait(ddf.shuffle("id", shuffle="tasks", shuffle_dtype=False).persist())
``` 

**Default**
```
18.4 s ± 241 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Minimal dtype**
```
19 s ± 247 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**No "_partitions"**
```
16.8 s ± 124 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I am happy to avoid the creation of "_partitions" (whenever possible) if it seems reasonable to others :)

TomAugspurger commented:
Thanks. I noticed that there's a decent amount of variability in the size of the hashed partitions

![rows](https://user-images.githubusercontent.com/1312546/80483490-ccb8b200-891b-11ea-9806-697491961c98.png)

Some of that is expected, since the `id` column isn't uniform. But there's perhaps more unevenness than I'd expect. For example, partition `214` has values from `id = [983, 982, 1046]`.

rjzamora commented:
> Some of that is expected, since the id column isn't uniform. But there's perhaps more unevenness than I'd expect. For example, partition 214 has values from id = [983, 982, 1046].

Good observation @TomAugspurger - We definitley do not get balanced partitions when the number of uniques in the `on` column(s) is not much larger than the number of output partitions.  In the example above, the number of output partitions is likely to be in the same ballpark as the number of unique values in `"id"`:

```python
print(ddf.npartitions, len(ddf["id"].unique()))
```
**Output**: `365 321` (Of course, the second number will vary)

For this reason, I think it makes sense that the output will be unbalanced.  You will see much better balance if you do something like:

```python
ddf = timeseries(start='2000-01-01', end='2000-12-31', partition_freq='1d', id_lam=100_000_000)
```

Perhaps we should add a note to the `DataFrame.shuffle` docstring that the index must have many uniques to produced balanced partitions?

rjzamora commented:
Note: I will need to revise the changes here a bit to alight with #6137

rjzamora commented:
Update:

- Removed `shuffle_dtype`.  Will always avoid creating `"_partitions"` when the user is passing in a column name (or list of column names) to `DataFrame.shuffle`.
- Added the `_simple_rearrange_by_column_tasks` code path for the case that the output partition count is small (less than `max_branch`).  The primary motivation is the case that the output partition count is small, but different from the input partition count. Unless I am misunderstanding, the current algorithm will first perform an `df.npartition`-to-`df.npartition` shuffle, and then repartition. A simpler algorithm can do this all in a single shuffle.

rjzamora commented on dask/dataframe/tests/test_merge_column_and_index.py:
```diff
@@ -147,7 +147,6 @@ def test_merge_known_to_unknown(
     # Assertions
     assert_eq(result, expected)
     assert_eq(result.divisions, tuple(None for _ in range(11)))
-    assert len(result.__dask_graph__()) >= 400
```
Do you know the motivation for these checks @TomAugspurger ? (or @jonmmease, who initially added them ?)

It seems that we are changing the graph size here, so the assertions are failing.

jonmmease replied:
I believe the graph size checks were added in cases where the merge algorithm should be able to avoid performing a full shuffle (it should be possible to perform the operation by just lining up the corresponding partitions).  The sizes were chosen empirically to be something smaller than the task graph that would be required to perform the same operation using a full shuffle.

Without understanding what this PR does, I'd say make sure that it's still the case that a full shuffle is not happening, and then bump up the number as needed.

rjzamora replied:
Great - Thanks for the info! I'll make sure we are still avoiding the full shuffle

rjzamora replied:
I just got a chance to look at these tests again. It seems that the assertion is making sure that the full shuffle **is** being performed (by checking that the graph is "large enough"), rather than the other way around. That is, the divisions are not known on one side, so the merge will require shuffling.  Since the changes in this PR result in a simpler shuffle for these cases, the graph is now smaller than "400" (more like 390).

@TomAugspurger - Does it make sense to change this check to `>=390` ? Or is it best to leave out the graph-size check moving forward?

TomAugspurger replied:
Mmm, let's just check >= 390 :)

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1177,6 +1177,62 @@ def repartition(
+            but index values will not be preserved.
+        compute: bool
+            Whether or not to trigger an immediate computation. Defaults to False.
+            Note, that even if you set ``compute=False``, an immediate computation
+            will still be triggered if ``divisions`` is ``None``.
```
Does `divisions` here refert to `self.divisions`? Or is this supposed to refer to a parameter?

rjzamora replied:
Ah - Sorry, this was a copy-paste mistake from the set_index description of `compute` - Since we are using hash values (rather than sorted order), there is no need to pass in divisions.

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -262,6 +270,24 @@ def shuffle(df, index, shuffle=None, npartitions=None, max_branch=32, compute=No
     set_partition
     shuffle_disk
     """
+    if shuffle == "tasks" and isinstance(index, (str, list)):
```
`index` is coming straight from users, so we should be a bit more forgiving on the list type. Perhaps use `pandas.api.types.is_list_like(index)`.

rjzamora replied:
Good point - I am now allowing the index to be "list_like", as long as it is not a dask collection.

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1177,6 +1177,62 @@ def repartition(
+        """ Rearrange DataFrame into new partitions
+
+        Uses hashing of `on` to map rows to output partitions. After this
+        operation, rows with the same index element(s) will be in the same
```
What does "index element(s)" mean here? Should this be "value of `on`"?

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -578,11 +679,17 @@ def rearrange_by_column_tasks(
     if npartitions is not None and npartitions != df.npartitions:
         token = tokenize(df2, npartitions)
-
+        repartition_get_name = "repartition-get-" + token
```
This line can be deleted?

rjzamora commented:
> IIUC, this hasn't changed the data model at all: there's no indication on the object that the dataframe is now partitioned by one or more columns. Should we? For index-based partitioning, we have .divisions, and known divisions are reflected in the repr. Or perhaps that doesn't make sense to add, since we don't know the hashed values like we know the divisions? I haven't thought through this fully.

Right - I do think it ultimately makes sense to expand the dask.dataframe data model to allow the `divisions` attribute to refer to be represented as a pd.DataFrame-like object so that the divisions can can correspond to one or more columns (including or not including the index).  With that said,  the shuffle method added here is not resulting in a "order"-based partitioning,  so the same concept of "divisions" is not really useful (unless we are interested in eventually storing information beyond lexicographical ordering in the `divisions` - which seems like a stretch to me)

TomAugspurger commented:
Yeah, agreed that we can't have a `.divisions`-like attribute for column shuffle-partitioned datasets. Just wondering if the fact that it's partitioned on that column's values should be reflected in the repr. But that can wait till later!

rjzamora commented:
> @rjzamora do you have strong thoughts on if we should / how to expose (hash-based) column partioning in the data model? I'd like to have an issue to collect discussion on this topic, but I'm still working through it in my head. If you don't have strong thoughts then I'll assign myself a task to write up an issue.

No strong thoughts from me, but I will be very happy to participate in a discussion :)...  Withouth giving it too much thought, I think we will probably want to allow the `divisions` to be represented by something like a `pd.DataFrame` object.  This way, the divisions can correspond to either an Index/MultiIndex, or an arbitrary set of columns. We may end up wanting to introduce a new `Divisions` class to organize the necessary attributes.  If the divisions correspond to hashed values rather than literal values, defining an attribute like `hash_partitioned=True` may do the trick.

rjzamora commented:
Looks like failures are unrelated.

TomAugspurger commented:
Yep, those were just fixed on master. I'll merge this and open an issue on the data model things.
